stdio_semihosting: Extend with RISC-V support

bergzand opened the pull request:
### Contribution description

RISC-V support semihosting in very similar way as the cortex-m
microcontrollers. The code calls a breakpoint instruction and the
attached debugger reads/writes registers and memory for stdio.

The RISC-V architecture doesn't support a call number with the EBREAK
instruction, to allow the debugger to detect a semihosting break point,
the EBREAK instruction is wrapped in a SLLI and SRAI instruction. These
use x0 as output register, making them NOP instructions.

One caveat when using this is that the RISC-V core traps the EBREAK
instruction with trap code 3 when no debugger is attached. Restarting
the application with the debugger attached avoids this.

This also piggybacks some enhancements to the fe310 and RISC-V makefiles to fix and expose the CPU_CORE_*, CPU_ARCH_* and CPU_MODEL_* to the compiler.

### Testing procedure

Test with `tests/sys_stdio_semihosting` on a RISC-V based board. The core might hang in trap number 3 (breakpoint) when initially attaching, restarting the core should solve this and it should work fine after this.

### Issues/PRs references

None

aabadie commented on cpu/fe310/Kconfig:
```diff
@@ -9,14 +9,14 @@ config CPU_ARCH_RISCV
     select HAS_ARCH_RISCV
 
-config CPU_CORE_RV32M
+config CPU_CORE_RV32I
```
Ok to change RV32M but for the fe310 the E31 core is a RV32IMAC.

aabadie replied:
Or maybe even simpler, `CPU_CORE_E31` in the case of the FE310.

aabadie commented:
I tested on hifive1b and it doesn't work (either when using JLink or OpenOCD as programmer). I'm able to make it work on an ARM board.
The README of the `tests/sys_stdio_semihosting` application should also be adapted to mention RISC-V in the list of supported architectures. For the moment, it stills only mention that an ARM based board can be used.

bergzand commented:
Rebased and fixed. This should now work on the hifive1b after enabling semihosting on the jlink with `monitor arm semihosting enable`
```
koen@morgen ~/dev/RIOT-review $ GDB=gdb  make -C tests/sys_stdio_semihosting/ BOARD=hifive1b term
make: Entering directory '/home/koen/dev/RIOT-review/tests/sys_stdio_semihosting'
/home/koen/dev/RIOT-review/dist/tools/jlink/jlink.sh debug /home/koen/dev/RIOT-review/tests/sys_stdio_semihosting/bin/hifive1b/tests_sys_stdio_semihosting.elf 
### Starting Debugging ###
SEGGER J-Link GDB Server V6.80c Command Line Version

JLinkARM.dll V6.80c (DLL compiled Jun 19 2020 17:23:30)

-----GDB Server start settings-----
GDBInit file:                  none
GDB Server Listening port:     3333
SWO raw output listening port: 2332
Terminal I/O port:             4444
Accept remote connection:      yes
Generate logfile:              off
Verify download:               off
Init regs on start:            off
Silent mode:                   on
Single run mode:               off
Target connection timeout:     0 ms
------J-Link related settings------
J-Link Host interface:         USB
J-Link script:                 none
J-Link settings file:          none
------Target related settings------
Target device:                 FE310
Target interface:              JTAG
Target interface speed:        2000kHz
Target endian:                 little

Reading symbols from /home/koen/dev/RIOT-review/tests/sys_stdio_semihosting/bin/hifive1b/tests_sys_stdio_semihosting.elf...
Remote debugging using :3333
Could not connect to J-Link.
Please check power, connection and settings.0x20010eca in _semihosting_raw (cmd=cmd@entry=5, args=args@entry=0x80000544 <main_stack+916>)
    at /home/koen/dev/RIOT-review/sys/stdio_semihosting/stdio_semihosting.c:71
71          __asm__(
(gdb) monitor arm semihosting enable
semihosting is enabled

(gdb) cont
Continuing.

> help
keep_alive() was not invoked in the 1000 ms timelimit. GDB alive packet not sent! (1096 ms). Workaround: increase "set remotetimeout" in GDB
help
Command              Description
---------------------------------------
reboot               Reboot the node
version              Prints current RIOT_VERSION
pm                   interact with layered PM subsystem
ps                   Prints information about running threads.
> ps
keep_alive() was not invoked in the 1000 ms timelimit. GDB alive packet not sent! (4273 ms). Workaround: increase "set remotetimeout" in GDB
ps
        pid | name                 | state    Q | pri | stack  ( used) ( free) | base addr  | current     
          1 | idle                 | pending  Q |  15 |    256 (  192) (   64) | 0x800000b0 | 0x800000f0 
          2 | main                 | running  Q |   7 |   1280 (  796) (  484) | 0x800001b0 | 0x800005f0 
            | SUM                  |            |     |   1536 (  988) (  548)
>
```

aabadie commented:
`monitor arm semihosting enable` this looks weird on a riscv. This doesn't work on my hifive1b, but semihosting is enabled with `monitor semihosting enable`. Unfortunately semihosting doesn't after when using jlink.
It works like a charm when using openocd as programmer though.

aabadie commented:
Please squash

aabadie reviewed (approved):
ACK

The problem I have locally seems related to JLink and not to the semihosting implementation itself: it works like a charm when using openocd.
I tried different version of JLink and all end up with the semihosting not working.

@bergzand, are you using the on-board programmer or an external JLink plugged on the JTAG pins ?

bergzand commented:
Rebased!

> @bergzand, are you using the on-board programmer or an external JLink plugged on the JTAG pins ?

The onboard one of the hifive1b

aabadie commented:
> The onboard one of the hifive1b

So I have no idea what's wrong with my setup...

bergzand commented:
Thanks for reviewing!
